accepted/engine-host-functions.md

```diff
@@ -0,0 +1,211 @@
+    let mem = caller.get_export("memory").unwrap().into_memory().unwrap();
+    println!(
+        "Hello {}!",
+        std::str::from_utf8(unsafe{ &mem.data_unchecked()[ptr..ptr + len] }).unwrap()
+    );
+})?;
+ 
+let module = Module::new(
+    &engine,
+    r#"
+    (module
+        (import "" "hello" (func $hello (param i32 i32)))
+        (memory (export "memory") 1)
+        (func (export "run")
+            i32.const 0
+            i32.const 5
+            call $hello
+        )
+        (data (i32.const 0) "world")
+    )
+    "#,
+)?;
+ 
+// Note: given that `Module` is `Sync + Send`, this could be on a request handling thread
+// As nothing is defined locally in the `Linker`, the engine definitions will
+// be used for satisfying imports
+let store = Store::new(module.engine());
+let linker = Linker::new(&store);
+ 
+let instance = linker.instantiate(&module)?;
+let run = instance
+    .get_export("run")
+    .unwrap()
+    .into_func()
+    .unwrap()
+    .get0::<()>();
+ 
+run()?;
+```
+ 
+## Changes to `wasmtime::Engine`
+ 
+This RFC proposes a `define_function` method be added to `Engine`:
+ 
+```rust
+/// Defines a function for the `Engine`.
+///
+/// Use `Store::get_engine_function` to get a `Func` representing the function.
+///
+/// A `Linker` will automatically default to functions on the associated `Engine`
+/// when resolving a function import that is not defined locally to the linker.
+pub fn define_function<Params, Results>(&self, module: &str, name: &str, func: impl IntoEngineFunc<Params, Results>) -> Result<()>;
+```
+
+Similar to `Func::wrap`, `define_function` will generically accept different `Fn` signatures to determine the WebAssembly type of the function. Instead of using `IntoFunc`, a different `IntoEngineFunc` trait will be used (see related section below).
+ 
+This method will internally create an `InstanceHandle` to represent the function.
+ 
+However, it will not associate the instance handle with a store; instead, the engine will own the associated instance handles and deallocate them when required (e.g. the `Engine` drops or an existing function definition is redefined).
+ 
+## Changes to `wasmtime::Store`
+ 
+To use `Engine` defined functions as imports for module instantiation or as `funcref` values, a `Func` representation is required.
+
+This proposal calls for adding the following methods to `Store`:
+ 
+```rust
+/// Gets a function from the `Engine` associated with this `Store`.
+///
+/// Returns `None` if the given function is not defined.
+pub fn get_engine_function(&self, module: &str, name: &str) -> Option<Func>;
+
+/// Get a reference to a context value of a particular type.
+///
+/// Returns `None` if the context value of the given type does not exist.
+pub fn get_context<T: Any>(&self) -> Option<&T>;
+ 
+/// Sets a context value.
+///
+/// Returns the previous context value of the same type if present.
+pub fn set_context<T: Any>(&mut self, value: T) -> Option<T>;
```

Comment:
Since `Store` is generally behind `Rc` and such I think we'll want to change this and `remove_context` to `&self` instead of `&mut self`.

Reply:
Agreed.  I'll remove the `&mut`.